### DISCUSSION TITLE
Adding form fields to a PDF

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#1405] [Looking for help] [answered]
Is there a method in PyMuPDF to add form fields to PDF

**Comment by JorjMcKie** [comment, marked as answer]
Sure - here is an example script: [widgettest.zip](https://github.com/pymupdf/PyMuPDF/files/7545026/widgettest.zip)

**Reply by meghanaviyyapu** [reply]
Thank you, I have extracted images from a PDF using PyMuPDF. The following image was not extracted.
![Capture](https://user-images.githubusercontent.com/36301306/141961907-5ab1ea04-a8b3-44cc-8628-b3791cb78c4e.PNG)
[ieee-format.pdf](https://github.com/pymupdf/PyMuPDF/files/7545240/ieee-format.pdf)
Can you let me know if there is a way to extract such images.

**Reply by JorjMcKie** [reply]
The reason: this is no image! It is a drawing.

You can extract drawings by `page.get_drawings()`. A list of dictionaries, which contain all the commands that result in drawings. This is very granular information - allowing you to regenerate the respective drawing on a new page (using the `Shape`  class).

You have to find out a way yourself, how to determine the page area (rectangle) that contains interesting graphics.

**Reply by meghanaviyyapu** [reply]
Thanks, do you know any ways to get only specific graphics?

**Reply by JorjMcKie** [reply]
I was wrong: this is an image - unexpectedly, xref=285

**Reply by JorjMcKie** [reply]
> Thanks, do you know any ways to get only specific graphics?

It all depends on you and what you know about the page.

**Reply by JorjMcKie** [reply]
If you have somehow figured out, which parts of the page interest you, then do a `page.get_pixmap(clip=rect)` with the interesting rectangle.

**Reply by meghanaviyyapu** [reply]
Is it possible to extract all the elements(drawings,paragraphs,images,lists,forms,headings) from one PDF and create a new PDF with the same appearance and the order in which the elements are present in the original PDF?

**Reply by JorjMcKie** [reply]
No

**Reply by JorjMcKie** [reply]
If you want all that, then why don't you simply insert that page in another PDF!

**Reply by meghanaviyyapu** [reply]
I want to tag all the elements in a PDF so I was thinking of extracting all elements from an existing PDF and create a new PDF with elements and their tags.

**Reply by JorjMcKie** [reply]
Ah, I see. Sometimes MS Word or LibreOffice do a good job when opening a PDF. Then you can work with the resulting office document ...

**Reply by meghanaviyyapu** [reply]
Is there a way to extract text and its  (x,y) co-ordinates?

**Comment by JorjMcKie** [comment]
> Is there a way to extract text and its (x,y) co-ordinates?

You don't know `page.get_text()`? For all method variants "words", "blocks", "dict", "rawdict" you get the full bbox coordinates.

**Reply by meghanaviyyapu** [reply]
Yeah I know that but instead of that I wanted to know if there is any other method.

**Comment by JorjMcKie** [comment]
well there also is `page.get_texttrace()`- comparable with "rawdict" but with even more information

**Reply by meghanaviyyapu** [reply]
![image](https://user-images.githubusercontent.com/36301306/142212712-06b260f2-5a14-4d46-8550-dc6ecb529469.png)
I am getting the above error when I used page.get_texttrace()

**Reply by JorjMcKie** [reply]
install more current version and lookup relevant documentation

**Reply by meghanaviyyapu** [reply]
Is there a way to highlight content in the PDF using xref values?
e.g.
![image](https://user-images.githubusercontent.com/36301306/143042652-9f55ef71-05c9-4c38-b45c-c7fd9d05f4eb.png)
In the above image, /S /P indicates that xref 61 value contains paragraph information. Is it possible to highlight the paragraph content in the PDF?
[Creating_PDFA.pdf](https://github.com/pymupdf/PyMuPDF/files/7588758/Creating_PDFA.pdf)

**Reply by JorjMcKie** [reply]
I don't know actually - as I wrote once, I have no experience in that area.
But in general, if you can figure out how to locate a rectangle (or quad) based on the information given by (or pointed to by) this xref, it is certainly no problem to create a highlight annotation.

**Reply by meghanaviyyapu** [reply]
Okay Thanks

**Reply by meghanaviyyapu** [reply]
Can you let me know based on what properties page.extract_text() extracts text from PDF without depending on content stream?

**Reply by JorjMcKie** [reply]
The correct method name is `page.get_text()`.
It reads out information stored in the `TextPage` object. This is a wrapper of MuPDF's `fz_stext_page`. When MuPDF code creation this textpage structure, it of course reads the page's `/Contents` objects for a PDF - or the corresponding information for other document types.

**Reply by meghanaviyyapu** [reply]
Using page.get_text(), PDF text is being rendered in ASCII format but when the text is extracted from the content stream(using xref of /Contents), all the contents are not in ASCII format. Is there a way to get the output identical to page.get_text() even when the text is extracted from the stream?

**Reply by JorjMcKie** [reply]
> Is there a way to get the output identical to page.get_text() ...

No:
1. The text is extracted as unicode, not just ASCII.
2. The `/Contents` object almost never contains text in directly readable form. Only the most simple fonts (e.g. Base-14) will be like this. Instead you will see glyph numbers (in 4-digit hex format). They are used to communicate with the font program and cannot be backtranslated to original text (by you or by me).

Didn't we discuss this already earlier in another post?

**Reply by meghanaviyyapu** [reply]
Yeah we had a discussion but I had a doubt how page.get_text() is extracting the content in ASCII or unicode format because as you mentioned backtranslation is not possible using the text present in /Contents object.
